src/Prelude/Linear.hs

```diff
@@ -74,3 +83,29 @@ maybe _ f (Just y) = f y
 -- arrow but we have a linear arrow
 forget :: (a ->. b) ->. a -> b
 forget f x = f x
+
+-- | Linearly typed replacement for the standard 'foldr' function, to allow
+-- linear consumption of a list.
+foldr :: (a ->. b ->. b) -> b ->. [a] ->. b
+foldr f z = \case
+  [] -> z
+  x:xs -> f x (foldr f z xs)
+
+-- | Replacement for the flip function with generalized multiplicities.
+flip :: (a -->.(p) b -->.(q) c) -->.(r) b -->.(q) a -->.(p) c
+flip f b a = f a b
+
+-- | Linearly typed replacement for the standard '(<$)' function.
+(<$) :: (Applicative f, Consumable b) => a ->. f b ->. f a
+a <$ fb = pure (flip lseq a) <*> fb
+
+-- | Linearly typed replacement for the standard '(<*)' function.
+(<*) :: (Applicative f, Consumable b) => f a ->. f b ->. f a
+fa <* fb = fmap (flip lseq) fa <*> fb
+
+-- | Linearly typed replacement for the standard 'foldM' function.
+foldM :: forall m a b. Monad m => (b ->. a ->. m b) -> b ->. [a] ->. m b
+foldM f z0 xs = foldr f' return xs z0
+  where
+    f' :: a ->. (b ->. m b) ->. b ->. m b
+    f' x k z = f z x >>= k
```

Comment:
I also wanted to add `foldM` to `Control.Monad.Linear` but then I was unsure where to put its dependency `foldr`.

Reply:
`Prelude.Linear.Internal.Simple`